wandb/data_types.py

```diff
@@ -749,6 +749,71 @@ def json_helper(val):
         return json_dict
 
 
+class PartitionedTable(Media):
+    """ PartitionedTable represents a table which is composed
+    by the union of multiple sub-tables. Currently, PartitionedTable
+    is designed to point to a directory within an artifact.
+    """
+
+    artifact_type = "partitioned-table"
+
+    def __init__(self, parts_path):
+        """
+        Args:
+            parts_path (str): path to a directory of tables in the artifact
+        """
+        super(PartitionedTable, self).__init__()
+        self.parts_path = parts_path
+        self._loaded_parts = []
+
+    def to_json(self, artifact):
+        json_obj = super(PartitionedTable, self).to_json(artifact)
+        json_obj["parts_path"] = self.parts_path
+        return json_obj
+
+    @classmethod
+    def from_json(cls, json_obj, source_artifact):
+        instance = cls(json_obj["parts_path"])
+        entries = source_artifact.manifest.get_entries_in_directory(
+            json_obj["parts_path"]
+        )
+        for entry in entries:
+            part = source_artifact.get(entry.path)
+            if isinstance(part, Table):
+                instance._add_part(part)
+        return instance
+
+    def materialize(self):
```

Comment:
1. I wonder if materialize is the right term for this? Imagine a huge set of tables, you wouldn't want to actually concatenate in memory. Rather, you'd have some interface that filter rows, stream rows, etc, that both types would implement.

2. If I call materialize before from_json, I get... an empty table? Seems like it should throw an error.

3. If I call from_json twice before materialize, I would get everything doubled. Again should be an error.

Reply:
1. I will convert this all to generators to make loops easy and minimize data load.
2+3. You cannot, since from_json is a class method

Reply:
I modified this so that instead we expose `.iterrows()` (the exact same method as pandas) which returns an Iterator (or more precisely a generator), which will loop through the rows with the same format as pandas. Moreover, added the same tech to the table. Under the hood, we only keep in memory the tables which are currently being iterated over.